File: examples/6_Visualization/visualization.cc
// ************************************************************************
// * This file is part of GGEMS.                                          *
// *                                                                      *
// * GGEMS is free software: you can redistribute it and/or modify        *
// * it under the terms of the GNU General Public License as published by *
// * the Free Software Foundation, either version 3 of the License, or    *
// * (at your option) any later version.                                  *
// *                                                                      *
// * GGEMS is distributed in the hope that it will be useful,             *
// * but WITHOUT ANY WARRANTY; without even the implied warranty of       *
// * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
// * GNU General Public License for more details.                         *
// *                                                                      *
// * You should have received a copy of the GNU General Public License    *
// * along with GGEMS.  If not, see <https://www.gnu.org/licenses/>.      *
// *                                                                      *
// ************************************************************************

/*!
  \file visualization.cc

  \brief Example of simulation with visualization

  \author Julien BERT <julien.bert@univ-brest.fr>
  \author Didier BENOIT <didier.benoit@inserm.fr>
  \author LaTIM, INSERM - U1101, Brest, FRANCE
  \version 1.0
  \date Monday October 25, 2021
*/

#include <sstream>
#include "GGEMS/global/GGEMSOpenCLManager.hh"
#include "GGEMS/graphics/GGEMSOpenGLManager.hh"
#include "GGEMS/global/GGEMS.hh"
#include "GGEMS/materials/GGEMSMaterialsDatabaseManager.hh"
#include "GGEMS/navigators/GGEMSVoxelizedPhantom.hh"
#include "GGEMS/navigators/GGEMSCTSystem.hh"
#include "GGEMS/physics/GGEMSRangeCutsManager.hh"
#include "GGEMS/physics/GGEMSProcessesManager.hh"
#include "GGEMS/sources/GGEMSXRaySource.hh"
#include "GGEMS/geometries/GGEMSVolumeCreatorManager.hh"
#include "GGEMS/geometries/GGEMSBox.hh"

#ifdef _WIN32
#include "GGEMS/tools/GGEMSWinGetOpt.hh"
#else
#include <getopt.h>
#endif

/*!
  \fn void PrintHelpAndQuit(std::string const& message, char const *p_executable)
  \param message - error message
  \param p_executable - name of the executable
  \brief print the help or the error of the program
*/
void PrintHelpAndQuit(std::string const& message, char const* exec)
{
  std::ostringstream oss(std::ostringstream::out);
  oss << message << std::endl;
  oss << std::endl;
  oss << "-->> 6 - Visualization Example <<--\n" << std::endl;
  oss << "Usage: " << exec << " [OPTIONS...]\n" << std::endl;
  oss << "[--help]                   Print the help to the terminal" << std::endl;
  oss << "[--verbose X]              Verbosity level" << std::endl;
  oss << "                           (X=0, default)" << std::endl;
  oss << std::endl;
  oss << "OpenGL params:" << std::endl;
  oss << "--------------" << std::endl;
  oss << "[--wdims X,Y]              Window dimensions" << std::endl;
  oss << "                           (800,800, by default)" << std::endl;
  oss << "[--msaa X]                 MSAA factor (1x, 2x, 4x or 8x)" << std::endl;
  oss << "                           (8, by default)" << std::endl;
  oss << "[--axis]                   Drawing axis on screen" << std::endl;
  oss << "[--wcolor X]               Window color" << std::endl;
  oss << "                           (black, by default)" << std::endl;
  oss << "Available colors:" << std::endl;
  oss << "                               * black" << std::endl;
  oss << "                               * blue" << std::endl;
  oss << "                               * lime" << std::endl;
  oss << "                               * cyan" << std::endl;
  oss << "                               * red" << std::endl;
  oss << "                               * magenta" << std::endl;
  oss << "                               * yellow" << std::endl;
  oss << "                               * white" << std::endl;
  oss << "                               * gray" << std::endl;
  oss << "                               * silver" << std::endl;
  oss << "                               * maroon" << std::endl;
  oss << "                               * olive" << std::endl;
  oss << "                               * green" << std::endl;
  oss << "                               * purple" << std::endl;
  oss << "                               * teal" << std::endl;
  oss << "                               * navy" << std::endl;
  oss << std::endl;
  oss << "Specific hardware selection:" << std::endl;
  oss << "----------------------------" << std::endl;
  oss << "[--device X]               Device Index:" << std::endl;
  oss << "                           (0, by default)" << std::endl;
  oss << std::endl;
  oss << "Simulation parameters:" << std::endl;
  oss << "----------------------" << std::endl;
  oss << "[--n-particles X]         Number of particles" << std::endl;
  oss << "                          (X=1000000, default)" << std::endl;
  oss << "[--seed X]                Seed of pseudo generator number" << std::endl;
  oss << "                          (X=777, default)" << std::endl;
  throw std::invalid_argument(oss.str());
}

/*!
  \fn void ParseCommandLine(std::string const& line_option, T* p_buffer)
  \tparam T - type of the array storing the option
  \param line_option - string from the command line
  \param p_buffer - buffer storing the commands
  \brief parse the command with comma
*/
template<typename T>
void ParseCommandLine(std::string const& line_option, T* p_buffer)
{
  std::istringstream iss(line_option);
  T* p = &p_buffer[0];
  while (iss >> *p++) if (iss.peek() == ',') iss.ignore();
}

/*!
  \fn int main(int argc, char** argv)
  \param argc - number of arguments
  \param argv - list of arguments
  \return status of program
  \brief main function of program
*/
int main(int argc, char** argv)
{
  try {
    // Verbosity level
    GGint verbosity_level = 0;

    // List of parameters
    GGint msaa = 8;
    GGint window_dims[] = {800, 800};
    std::string window_color = "black";
    GGsize number_of_particles = 1000000;
    GGsize device_index = 0;
    GGuint seed = 777;
    static GGint is_axis = 0;

    // Loop while there is an argument
    GGint counter(0);
    while (1) {
      // Declaring a structure of the options
      GGint option_index = 0;
      static struct option sLongOptions[] = {
        {"verbose", required_argument, 0, 'v'},
        {"help", no_argument, 0, 'h'},
        {"msaa", required_argument, 0, 'm'},
        {"wdims", required_argument, 0, 'w'},
        {"wcolor", required_argument, 0, 'c'},
        {"device", required_argument, 0, 'd'},
        {"seed", required_argument, 0, 's'},
        {"n-particles", required_argument, 0, 'n'},
        {"axis", no_argument, &is_axis, 1},
      };

      // Getting the options
      counter = getopt_long(argc, argv, "hv:m:w:c:d:s:n:", sLongOptions, &option_index);

      // Exit the loop if -1
      if (counter == -1) break;

      // Analyzing each option
      switch (counter) {
        case 0: {
          // If this option set a flag, do nothing else now
          if (sLongOptions[option_index].flag != 0) break;
          break;
        }
        case 'v': {
          ParseCommandLine(optarg, &verbosity_level);
          break;
        }
        case 'h': {
          PrintHelpAndQuit("Printing the help", argv[0]);
          break;
        }
        case 'm': {
          ParseCommandLine(optarg, &msaa);
          break;
        }
        case 'w': {
          ParseCommandLine(optarg, &window_dims[0]);
          break;
        }
        case 'c': {
          window_color = optarg;
          break;
        }
        case 'd': {
          ParseCommandLine(optarg, &device_index);
          break;
        }
        case 's': {
          ParseCommandLine(optarg, &seed);
          break;
        }
        case 'n': {
          ParseCommandLine(optarg, &number_of_particles);
          break;
        }
        default: {
          PrintHelpAndQuit("Out of switch options!!!", argv[0]);
          break;
        }
      }
    }

    // Setting verbosity
    GGcout.SetVerbosity(verbosity_level);
    GGcerr.SetVerbosity(verbosity_level);
    GGwarn.SetVerbosity(verbosity_level);

    // Initialization of singletons
    GGEMSOpenCLManager& opencl_manager = GGEMSOpenCLManager::GetInstance();
    GGEMSOpenGLManager& opengl_manager = GGEMSOpenGLManager::GetInstance();
    GGEMSMaterialsDatabaseManager& material_manager = GGEMSMaterialsDatabaseManager::GetInstance();
    GGEMSVolumeCreatorManager& volume_creator_manager = GGEMSVolumeCreatorManager::GetInstance();
    GGEMSProcessesManager& processes_manager = GGEMSProcessesManager::GetInstance();
    GGEMSRangeCutsManager& range_cuts_manager = GGEMSRangeCutsManager::GetInstance();

    // Visualization params
    opengl_manager.SetMSAA(msaa);
    if (is_axis) opengl_manager.SetDrawAxis(true);
    opengl_manager.SetWindowDimensions(window_dims[0], window_dims[1]);
    opengl_manager.SetBackgroundColor(window_color);
    opengl_manager.SetImageOutput("data/axis");
    opengl_manager.SetWorldSize(2.0f*m,2.0f*m,2.0f*m);
    opengl_manager.Initialize();
    // XXX.SetParticleColor("gamma","red") ...
    // XXX.SetParticles(10000) // Max: 1000000!!!

    // OpenCL params
    opencl_manager.DeviceToActivate(device_index);

    // Enter material database
    material_manager.SetMaterialsDatabase("data/materials.txt");

    // Initializing a global voxelized volume
    volume_creator_manager.SetVolumeDimensions(120, 120, 120);
    volume_creator_manager.SetElementSizes(0.1f, 0.1f, 0.1f, "mm");
    volume_creator_manager.SetOutputImageFilename("data/phantom.mhd");
    volume_creator_manager.SetRangeToMaterialDataFilename("data/range_phantom.txt");
    volume_creator_manager.SetMaterial("Air");
    volume_creator_manager.SetDataType("MET_INT");
    volume_creator_manager.Initialize();

    // Creating a box
    GGEMSBox* box_phantom = new GGEMSBox(10.0f, 10.0f, 10.0f, "mm");
    box_phantom->SetPosition(0.0f, 0.0f, 0.0f, "mm");
    box_phantom->SetLabelValue(1);
    box_phantom->SetMaterial("Water");
    box_phantom->Initialize();
    box_phantom->Draw();
    delete box_phantom;

    // Writing volume
    volume_creator_manager.Write();

    // Phantoms and systems
    GGEMSVoxelizedPhantom phantom("phantom");
    phantom.SetPhantomFile("data/phantom.mhd", "data/range_phantom.txt");
    phantom.SetRotation(0.0f, 0.0f, 0.0f, "deg");
    phantom.SetPosition(0.0f, 0.0f, 0.0f, "mm");
    phantom.SetVisible(true);
    // phantom.SetMaterialColor("Water", false);
    // phantom.SetMaterialColor("Water", 0.34, 0.4, 0.2, true);

    GGEMSCTSystem ct_detector("Stellar");
    ct_detector.SetCTSystemType("curved");
    ct_detector.SetNumberOfModules(1, 46);
    ct_detector.SetNumberOfDetectionElementsInsideModule(64, 16, 1);
    ct_detector.SetSizeOfDetectionElements(0.6f, 0.6f, 0.6f, "mm");
    ct_detector.SetMaterialName("GOS");
    ct_detector.SetSourceDetectorDistance(1085.6f, "mm");
    ct_detector.SetSourceIsocenterDistance(595.0f, "mm");
    ct_detector.SetRotation(0.0f, 0.0f, 0.0f, "deg");
    ct_detector.SetThreshold(10.0f, "keV");
    ct_detector.StoreOutput("data/projection");
    ct_detector.StoreScatter(true);
    ct_detector.SetVisible(true);
    //ct_detector.SetColor("red");

    // Physics
    processes_manager.AddProcess("Compton", "gamma", "all");
    processes_manager.AddProcess("Photoelectric", "gamma", "all");
    processes_manager.AddProcess("Rayleigh", "gamma", "all");

    // Optional options, the following are by default
    processes_manager.SetCrossSectionTableNumberOfBins(220);
    processes_manager.SetCrossSectionTableMinimumEnergy(1.0f, "keV");
    processes_manager.SetCrossSectionTableMaximumEnergy(1.0f, "MeV");

    // Cuts, by default but are 1 um
    range_cuts_manager.SetLengthCut("all", "gamma", 0.1f, "mm");

    // Source
    GGEMSXRaySource point_source("point_source");
    point_source.SetSourceParticleType("gamma");
    point_source.SetNumberOfParticles(number_of_particles);
    point_source.SetPosition(-595.0f, 0.0f, 0.0f, "mm");
    point_source.SetRotation(0.0f, 0.0f, 0.0f, "deg");
    point_source.SetBeamAperture(12.5f, "deg");
    point_source.SetFocalSpotSize(0.0f, 0.0f, 0.0f, "mm");
    point_source.SetPolyenergy("data/spectrum_120kVp_2mmAl.dat");

    // GGEMS simulation
    GGEMS ggems;
    ggems.SetOpenCLVerbose(true);
    ggems.SetMaterialDatabaseVerbose(false);
    ggems.SetNavigatorVerbose(false);
    ggems.SetSourceVerbose(true);
    ggems.SetMemoryRAMVerbose(true);
    ggems.SetProcessVerbose(true);
    ggems.SetRangeCutsVerbose(true);
    ggems.SetRandomVerbose(true);
    ggems.SetProfilingVerbose(true);
    ggems.SetTrackingVerbose(false, 0);

    // Initializing the GGEMS simulation
    ggems.Initialize(seed);

    opengl_manager.Display();

    // Start GGEMS simulation
    // ggems.Run();
  }
  catch (std::exception& e) {
    std::cerr << e.what() << std::endl;
    // Exit safely
    GGEMSOpenCLManager::GetInstance().Clean();
  }
  catch (...) {
    std::cerr << "Unknown exception!!!" << std::endl;
    // Exit safely
    GGEMSOpenCLManager::GetInstance().Clean();
  }

  // Exit safely
  GGEMSOpenCLManager::GetInstance().Clean();
}

File: examples/6_Visualization/visualization.py
# ************************************************************************
# * This file is part of GGEMS.                                          *
# *                                                                      *
# * GGEMS is free software: you can redistribute it and/or modify        *
# * it under the terms of the GNU General Public License as published by *
# * the Free Software Foundation, either version 3 of the License, or    *
# * (at your option) any later version.                                  *
# *                                                                      *
# * GGEMS is distributed in the hope that it will be useful,             *
# * but WITHOUT ANY WARRANTY; without even the implied warranty of       *
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
# * GNU General Public License for more details.                         *
# *                                                                      *
# * You should have received a copy of the GNU General Public License    *
# * along with GGEMS.  If not, see <https://www.gnu.org/licenses/>.      *
# *                                                                      *
# ************************************************************************

import argparse
from ggems import *

# ------------------------------------------------------------------------------
# Read arguments
parser = argparse.ArgumentParser(
  prog='visualization.py',
  description='''-->> 6 - OpenGL Visualization Example <<--''',
  epilog='''''',
  formatter_class=argparse.ArgumentDefaultsHelpFormatter)

parser.add_argument('-v', '--verbose', required=False, type=int, default=0, help='Set level of verbosity')
parser.add_argument('-e', '--nogl', required=False, action='store_false', help='Disable OpenGL')
parser.add_argument('-w', '--wdims', required=False, default=[800, 800], type=int, nargs=2, help='OpenGL window dimensions')
parser.add_argument('-m', '--msaa', required=False, type=int, default=8, help='MSAA factor (1x, 2x, 4x or 8x)')
parser.add_argument('-a', '--axis', required=False, action='store_true', help='Drawing axis in OpenGL window')
parser.add_argument('-p', '--nparticlesgl', required=False, type=int, default=256, help='Number of displayed primary particles on OpenGL window (max: 65536)')
parser.add_argument('-b', '--drawgeom', required=False, action='store_true', help='Draw geometry only on OpenGL window')
parser.add_argument('-c', '--wcolor', required=False, type=str, default='black', help='Background color of OpenGL window')
parser.add_argument('-d', '--device', required=False, type=str, default='0', help="OpenCL device running visualization")
parser.add_argument('-n', '--nparticles', required=False, type=int, default=1000000, help="Number of particles")
parser.add_argument('-s', '--seed', required=False, type=int, default=777, help="Seed of pseudo generator number")

args = parser.parse_args()

# Getting arguments
verbosity_level = args.verbose
seed = args.seed
device = args.device
number_of_particles = args.nparticles
number_of_displayed_particles = args.nparticlesgl
is_axis = args.axis
msaa = args.msaa
window_color = args.wcolor
window_dims = args.wdims
is_draw_geom = args.drawgeom
is_gl = args.nogl

# ------------------------------------------------------------------------------
# STEP 0: Level of verbosity during computation
GGEMSVerbosity(verbosity_level)

# ------------------------------------------------------------------------------
# STEP 1: Calling C++ singleton
opencl_manager = GGEMSOpenCLManager()
materials_database_manager = GGEMSMaterialsDatabaseManager()
processes_manager = GGEMSProcessesManager()
range_cuts_manager = GGEMSRangeCutsManager()
volume_creator_manager = GGEMSVolumeCreatorManager()

# ------------------------------------------------------------------------------
# STEP 2: Params for visualization
if is_gl:
  opengl_manager = GGEMSOpenGLManager()
  opengl_manager.set_window_dimensions(window_dims[0], window_dims[1])
  opengl_manager.set_msaa(msaa)
  opengl_manager.set_background_color(window_color)
  opengl_manager.set_draw_axis(is_axis)
  opengl_manager.set_world_size(3.0, 3.0, 3.0, 'm')
  opengl_manager.set_image_output('data/axis')
  opengl_manager.set_displayed_particles(number_of_displayed_particles)
  opengl_manager.set_particle_color('gamma', 152, 251, 152)
  # opengl_manager.set_particle_color('gamma', color_name='red') # Using registered color
  opengl_manager.initialize()

# ------------------------------------------------------------------------------
# STEP 3: Choosing an OpenCL device
opencl_manager.set_device_to_activate(device)

# ------------------------------------------------------------------------------
# STEP 4: Setting GGEMS materials
materials_database_manager.set_materials('data/materials.txt')

# ------------------------------------------------------------------------------
# STEP 5: Phantoms and systems

# Generating phantom
volume_creator_manager.set_dimensions(120, 120, 120)
volume_creator_manager.set_element_sizes(1.0, 1.0, 1.0, 'mm')
volume_creator_manager.set_output('data/phantom.mhd')
volume_creator_manager.set_range_output('data/range_phantom.txt')
volume_creator_manager.set_material('Air')
volume_creator_manager.set_data_type('MET_INT')
volume_creator_manager.initialize()

box_phantom = GGEMSBox(80.0, 80.0, 80.0, 'mm')
box_phantom.set_position(0.0, 0.0, 0.0, 'mm')
box_phantom.set_label_value(1)
box_phantom.set_material('Water')
box_phantom.initialize()
box_phantom.draw()
box_phantom.delete()

volume_creator_manager.write()

# Loading phantom in GGEMS
phantom = GGEMSVoxelizedPhantom('phantom')
phantom.set_phantom('data/phantom.mhd', 'data/range_phantom.txt')
phantom.set_rotation(0.0, 0.0, 0.0, 'deg')
phantom.set_position(0.0, 0.0, 0.0, 'mm')
phantom.set_visible(True)
phantom.set_material_visible('Air', True)
phantom.set_material_color('Water', color_name='blue') # Uncomment for automatic color

cbct_detector = GGEMSCTSystem('custom')
cbct_detector.set_ct_type('flat')
cbct_detector.set_number_of_modules(1, 1)
cbct_detector.set_number_of_detection_elements(400, 400, 1)
cbct_detector.set_size_of_detection_elements(1.0, 1.0, 10.0, 'mm')
cbct_detector.set_material('GOS')
cbct_detector.set_source_detector_distance(1500.5, 'mm') # Center of inside detector, adding half of detector (= SDD surface + 10.0/2 mm half of depth)
cbct_detector.set_source_isocenter_distance(900.0, 'mm')
cbct_detector.set_rotation(0.0, 0.0, 0.0, 'deg')
cbct_detector.set_global_system_position(0.0, 0.0, 0.0, 'mm');
cbct_detector.set_threshold(10.0, 'keV')
cbct_detector.save('data/projection')
cbct_detector.store_scatter(True)
cbct_detector.set_visible(True)
cbct_detector.set_material_color('GOS', 255, 0, 0) # Custom color using RGB
#cbct_detector.set_material_color('GOS', color_name='red') # Using registered color

cbct_detector2 = GGEMSCTSystem('custom2')
cbct_detector2.set_ct_type('flat')
cbct_detector2.set_number_of_modules(1, 1)
cbct_detector2.set_number_of_detection_elements(400, 400, 1)
cbct_detector2.set_size_of_detection_elements(1.0, 1.0, 10.0, 'mm')
cbct_detector2.set_material('Silicon')
cbct_detector2.set_source_detector_distance(1605.0, 'mm') # Center of inside detector, adding half of detector (= SDD surface + 10.0/2 mm half of depth)
cbct_detector2.set_source_isocenter_distance(1200.0, 'mm')
cbct_detector2.set_rotation(0.0, 0.0, 90.0, 'deg')
cbct_detector2.set_global_system_position(0.0, 0.0, 0.0, 'mm');
cbct_detector2.set_threshold(10.0, 'keV')
cbct_detector2.save('data/projection2')
cbct_detector2.store_scatter(True)
cbct_detector2.set_visible(True)

# ------------------------------------------------------------------------------
# STEP 6: Physics
processes_manager.add_process('Compton', 'gamma', 'all')
processes_manager.add_process('Photoelectric', 'gamma', 'all')
processes_manager.add_process('Rayleigh', 'gamma', 'all')

# Optional options, the following are by default
processes_manager.set_cross_section_table_number_of_bins(220)
processes_manager.set_cross_section_table_energy_min(1.0, 'keV')
processes_manager.set_cross_section_table_energy_max(1.0, 'MeV')

# ------------------------------------------------------------------------------
# STEP 7: Cuts, by default but are 1 um
range_cuts_manager.set_cut('gamma', 0.1, 'mm', 'all')

# ------------------------------------------------------------------------------
# STEP 8: Source
point_source = GGEMSXRaySource('point_source')
point_source.set_source_particle_type('gamma')
point_source.set_number_of_particles(number_of_particles)
point_source.set_position(-900.0, 0.0, 0.0, 'mm')
point_source.set_rotation(0.0, 0.0, 0.0, 'deg')
point_source.set_beam_aperture(12.5, 'deg')
point_source.set_focal_spot_size(0.2, 0.6, 0.0, 'mm')
point_source.set_polyenergy('data/spectrum_120kVp_2mmAl.dat')

point_source2 = GGEMSXRaySource('point_source2')
point_source2.set_source_particle_type('gamma')
point_source2.set_number_of_particles(number_of_particles)
point_source2.set_position(-1200.0, 0.0, 0.0, 'mm')
point_source2.set_rotation(0.0, 0.0, 90.0, 'deg')
point_source2.set_beam_aperture(8.5, 'deg')
point_source2.set_focal_spot_size(0.2, 0.6, 0.0, 'mm')
point_source2.set_polyenergy('data/spectrum_120kVp_2mmAl.dat')

# ------------------------------------------------------------------------------
# STEP 9: GGEMS simulation
ggems = GGEMS()
ggems.opencl_verbose(True)
ggems.material_database_verbose(False)
ggems.navigator_verbose(False)
ggems.source_verbose(True)
ggems.memory_verbose(True)
ggems.process_verbose(True)
ggems.range_cuts_verbose(True)
ggems.random_verbose(True)
ggems.profiling_verbose(True)
ggems.tracking_verbose(False, 0)

# Initializing the GGEMS simulation
ggems.initialize(seed)

if is_draw_geom and is_gl: # Draw only geometry and do not run GGEMS
  opengl_manager.display()
else: # Running GGEMS and draw particles
  ggems.run()

# ------------------------------------------------------------------------------
# STEP 10: Exit safely
clean_safely()
exit()
